Installation issues on `R 4.0` and `R 3.6`

Hi, thanks for the great package.

Our CI builds on R 3.6 are [failing](https://github.com/easystats/modelbased/actions/runs/9328926165/job/25680713253) because compilation seems to fail for this R version:

```r
  ../inst/include/quickjsr/JSValue_to_SEXP.hpp:72:37: note: candidate: ‘quickjsr::JSValue_to_SEXP_vector(JSContext*, const JSValue&)::<lambda(int&&)>’
     72 |                                     [len](auto&& vec) { return vec.size() == len; });
        |                                     ^
  ../inst/include/quickjsr/JSValue_to_SEXP.hpp:72:37: note:   no known conversion for argument 1 from ‘std::vector<double>’ to ‘int&&’
  make: *** [/opt/R/3.6.3/lib/R/etc/Makeconf:177: quickjsr.o] Error 1
  ERROR: compilation failed for package ‘QuickJSR’
  * removing ‘/tmp/Rtmpg2EnHZ/pkg-lib12f411d6546/QuickJSR’
  ```
  
  Is this expected? If yes, can you please set the minimum needed R version in `DESCRIPTION`? Thanks.

It looks like the package is being compiled using C++11, while it needs c++14 features:

```
  g++ -std=gnu++11 -I"/opt/R/3.6.3/lib/R/include" -DNDEBUG -I"../inst/include/" -I"../inst/include/quickjs" -D_GNU_SOURCE -DCONFIG_BIGNUM -DCONFIG_VERSION=\"2024-01-13\"  -I/usr/local/include  -fpic  -g -O2  -Wall -pedantic -c quickjsr.cpp -o quickjsr.o
  In file included from ../inst/include/quickjsr/JSValue_to_Cpp.hpp:4,
                   from ../inst/include/quickjsr/JSValue_to_SEXP.hpp:6,
                   from ../inst/include/quickjsr/SEXP_to_JSValue.hpp:5,
                   from ../inst/include/quickjsr.hpp:4,
                   from quickjsr.cpp:4:
  ../inst/include/quickjsr/type_traits.hpp:15:32: error: ‘decay_t’ in namespace ‘std’ does not name a template type; did you mean ‘decay’?
     15 |     using type = typename std::decay_t<T>;
        |                                ^~~~~~~
        |                                decay
```

I don't set a c++ standard in the package, since CRAN gives a NOTE for anything other than the default c++17

--- a/.github/workflows/R-CMD-check.yaml
+++ b/.github/workflows/R-CMD-check.yaml
@@ -29,6 +29,7 @@ jobs:
           - {os: ubuntu-latest,   r: 'devel', http-user-agent: 'release'}
           - {os: ubuntu-latest,   r: 'release'}
           - {os: ubuntu-latest,   r: 'oldrel'}
+          - {os: ubuntu-latest,   r: '3.6'}
 
     env:
       GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}

--- a/inst/include/quickjsr/JSValue_to_Cpp.hpp
+++ b/inst/include/quickjsr/JSValue_to_Cpp.hpp
@@ -44,7 +44,7 @@ namespace quickjsr {
     return result == "true" ? "TRUE" : result == "false" ? "FALSE" : result;
   }
 
-  template <typename T, std::enable_if_t<is_std_vector<T>::value>* = nullptr>
+  template <typename T, typename std::enable_if<is_std_vector<T>::value>::type* = nullptr>
   T JSValue_to_Cpp(JSContext* ctx, JSValue val) {
     T res;
     uint32_t len;

--- a/inst/include/quickjsr/JSValue_to_SEXP.hpp
+++ b/inst/include/quickjsr/JSValue_to_SEXP.hpp
@@ -69,7 +69,7 @@ SEXP JSValue_to_SEXP_vector(JSContext* ctx, const JSValue& val) {
       // Check that the inner vectors are all the same length
       size_t len = res[0].size();
       bool allSameSize = std::all_of(res.begin() + 1, res.end(),
-                                    [len](auto&& vec) { return vec.size() == len; });
+                                    [len](const std::vector<double>& vec) { return vec.size() == len; });
       if (allSameSize) {
         cpp11::writable::doubles_matrix<cpp11::by_column> out(res.size(), len);
         for (size_t i = 0; i < res.size(); i++) {

--- a/inst/include/quickjsr/type_traits.hpp
+++ b/inst/include/quickjsr/type_traits.hpp
@@ -12,18 +12,18 @@ namespace quickjsr {
 
   template <typename T, typename = void>
   struct value_type {
-    using type = typename std::decay_t<T>;
+    using type = typename std::decay<T>::type;
   };
 
   template <typename T>
-  struct value_type<T, std::enable_if_t<is_std_vector<T>::value>> {
-    using type = typename std::decay_t<T>::value_type;
+  struct value_type<T, typename std::enable_if<is_std_vector<T>::value>::type> {
+    using type = typename std::decay<T>::type::value_type;
   };
 
   template <typename T>
   using value_type_t = typename value_type<T>::type;
 
   template <typename GoalT, typename InT>
-  using enable_if_type_t = std::enable_if_t<std::is_same<GoalT, InT>::value>;
+  using enable_if_type_t = typename std::enable_if<std::is_same<GoalT, InT>::value>::type;
 } // namespace quickjsr
 #endif

--- a/inst/tinytest/test_JSContext.R
+++ b/inst/tinytest/test_JSContext.R
@@ -16,12 +16,12 @@ expect_equal(jsc$call("mult_test", 10, 15), 150)
 
 # Test that R functions can be passed and evaluated in JS
 jsc$source(code = "function fun_test(f, x, y) { return f(x, y); }")
-expect_equal(jsc$call("fun_test", \(x, y){ x + y }, 1, 2), 3)
+expect_equal(jsc$call("fun_test", function(x, y){ x + y }, 1, 2), 3)
 
 # Test that closures/captures work
 a <- 3
-expect_equal(jsc$call("fun_test", \(x, y){ (x + y) * a }, 1, 2), 9)
-expect_equal(jsc$call("fun_test", \(x, y){ paste(x, y) }, "a", "b"), "a b")
+expect_equal(jsc$call("fun_test", function(x, y){ (x + y) * a }, 1, 2), 9)
+expect_equal(jsc$call("fun_test", function(x, y){ paste(x, y) }, "a", "b"), "a b")
 
 # Test that R environments can be passed to JS and values accessed
 jsc$source(code = "function env_test(env) { return env.a + env.b; }")

--- a/inst/tinytest/test_to_json_date.R
+++ b/inst/tinytest/test_to_json_date.R
@@ -1,3 +1,6 @@
+# Fails on 3.6 CI, but can't be replicated locally
+exit_if_not(R.version$major  > "3")
+
 object <- as.Date("1985-06-18");
 
 expect_equal(to_json(object), "[\"1985-06-18T00:00:00.000Z\"]");

